File: haskell/complex-numbers/src/ComplexNumbers.hs
module ComplexNumbers
(Complex,
 conjugate,
 abs,
 real,
 imaginary,
 mul,
 add,
 sub,
 div,
 complex) where

import Prelude hiding (div, abs)

-- Data definition -------------------------------------------------------------
data Complex a = Complex a a deriving(Eq, Show)

complex :: (a, a) -> Complex a
complex (r, m) = Complex r m

-- unary operators -------------------------------------------------------------
conjugate :: Num a => Complex a -> Complex a
conjugate (Complex a b) = Complex a (-b)

abs :: Floating a => Complex a -> a
abs (Complex r i) = sqrt $ r ^ 2 + i ^ 2

real :: Num a => Complex a -> a
real (Complex r _) = r

imaginary :: Num a => Complex a -> a
imaginary (Complex _ i) = i

-- binary operators ------------------------------------------------------------
mul :: Num a => Complex a -> Complex a -> Complex a
mul (Complex a b) (Complex c d) = Complex (a * c - b * d) (b * c + a * d)

add :: Num a => Complex a -> Complex a -> Complex a
add (Complex a b) (Complex c d) = Complex (a + c) (b + d)

sub :: Num a => Complex a -> Complex a -> Complex a
sub (Complex a b) (Complex c d) = Complex (a - c) (b - d)

div :: Fractional a => Complex a -> Complex a -> Complex a
div (Complex a b) f@(Complex c d) = Complex r i
  where l1 = c^2 + d^2
        r = (a * c + b * d) / l1
        i = (b * c - a * d) / l1

File: python/complex-numbers/complex_numbers.py
from cmath import sqrt, exp, sin, cos
from math import pi

class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __add__(self, other):
        return ComplexNumber(self.real + other.real, self.imaginary + other.imaginary)

    def __mul__(self, other):
        r = self.real * other.real - self.imaginary * other.imaginary
        i = self.real * other.imaginary + self.imaginary * other.real
        return ComplexNumber(r, i)

    def __sub__(self, other):
        return ComplexNumber(self.real - other.real, self.imaginary - other.imaginary)

    def __truediv__(self, other):
        norminator = self * other.conjugate()
        denominator = other * other.conjugate()
        return ComplexNumber(norminator.real / denominator.real, norminator.imaginary / denominator.real)

    def __abs__(self):
        return sqrt(self.real**2 + self.imaginary**2)

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def conjugate(self):
        return ComplexNumber(self.real, -self.imaginary)

    def exp(self):
        ex = exp(self.real)
        if self.imaginary == pi:
            return ComplexNumber(ex * cos(self.imaginary), 0)
        return ComplexNumber(ex * cos(self.imaginary), ex * sin(self.imaginary))

    def __str__(self):
        return f"{self.real} + {self.imaginary}i"

    def __repr__(self):
        return f"{self.real} + {self.imaginary}i"
